bin/propolis-server/src/lib/stats.rs

```diff
@@ -201,34 +208,175 @@ pub async fn register_server_metrics(
 #[cfg(all(not(test), target_os = "illumos"))]
 async fn setup_kstat_tracking(
     log: &Logger,
+    nics: InstanceNetworkInterfaces,
     virtual_machine: VirtualMachine,
 ) -> Option<KstatSampler> {
-    let kstat_limit =
-        usize::try_from(virtual_machine.n_vcpus() * KSTAT_LIMIT_PER_VCPU)
-            .unwrap();
+    let n_interfaces = nics.interface_ids.len() as u32;
+    let kstat_limit = usize::try_from(
+        (virtual_machine.n_vcpus() * KSTAT_LIMIT_PER_VCPU)
+            + (n_interfaces * KSTAT_LIMIT_PER_VCPU),
+    )
+    .unwrap();
     match KstatSampler::with_sample_limit(log, kstat_limit) {
         Ok(sampler) => {
```

Comment:
This might be a little less tangled if it went something along the lines of:
```rust
let sampler = KstatSampler::with_sample_limit(log, kstat_limit).map_err(|e| {
    slog::error!(
        log,
        "failed to create KstatSampler, \
        vCPU and link stats will be unavailable";
        "error" => ?e,
    );
    e
}).ok()?;
let vcpu_details =
// ...
```